typing/ctype.ml

```diff
@@ -1245,10 +1354,35 @@ let rec copy ?partial ?keep_names ?scope copy_scope ty =
               For_copy.redirect_desc copy_scope more
                 (Tsubst(more', Some t));
               (* Return a new copy *)
+              let row = match row_name row with
+                | Some (p, tl) ->
+                    set_row_name row (Some (Path.subst id_map p, tl))
+                | None -> row
+              in
               Tvariant (copy_row copy true row keep more')
           end
       | Tobject (ty1, _) when partial <> None ->
           Tobject (copy ty1, ref None)
+      | Tobject (ty, {contents = Some (p, tl)}) ->
+          let p = Path.subst id_map p in
+          Tobject (copy ty, ref (Some (p, List.map copy tl)))
+      | Tfunctor (lbl, id, {pack_path = p; pack_cstrs}, ty) ->
+          let fl = List.map (fun (li, ty) -> (li, copy ty)) pack_cstrs in
+          let p = Path.subst id_map p in
+          let id' = Ident.Unscoped.refresh id in
+          let p_id' = Path.Pident (Ident.of_unscoped id') in
+          let id = Ident.of_unscoped id in
+          let id_map = (id, p_id')
+                :: List.filter (fun (i, _) -> not (Ident.same i id)) id_map in
+          let tyset =
+            type_subexpressions_with_occurrences (List.map fst id_map) ty in
```

Comment:
The logic here will traverse all subexpressions below a `Tfunctor`, each type a `Tfunctor` is encountered during the copy. It would be more efficient, on the top-most `Tfunctor` encountered, to collect the smallest set of nodes that contain its variable *or* a variable of Tfunctor itself present in that set. Then we would not need more traversals on Tfunctor children of this node.

This would make the code more complex, so we are leaving it for later.